### DISCUSSION TITLE
MagickCMD

### DISCUSSION THREAD

**Discussion by Stormwalker** [ImageMagick/ImageMagick#4685] [Development] [answered]
Hi, this is probably a dumb question, but I'm trying to find MagickCMD.cpp as mentioned in the docs for the ImageMagickObject and I can't seem to locate it?

Many thanks!

**Comment by dlemstra** [comment]
I suspect that file got removed and the documentation did not get updated.

**Reply by Stormwalker** [reply]
Hi Dirk,

Thanks for that. The reason I ask is that I work with a scripting language implemented in C and when I use the Function Description from the type library to build the IDispatch::Invoke arguments (e.g for the Identify method) it's not behaving as I expected.

e.g. The Identify method returns a FUNCDESC with cParams of 1, and a cOptParams of -1, which I believe means that it should take a single SafeArray containing the params according to the FUNCDESC docs.  However, the Perform method inside the ImageMagickObject treats the last array it finds as an output argument, which is my input array, so it fails  (If I ignore the FUNCDESC and just pass the arguments individually then it works because Perform is written to handle this case too)

So I wanted to see an example of how you guys are calling it from C/C++ to see if my expectations are incorrect - I don't suppose you would know where I could find another such example if MagickCMD isn't available?

Thanks for your help.

**Comment by dlemstra** [comment, marked as answer]
I don't think you should be using `ImageMagickObject` anymore. We haven't updated that in ages and I am even thinking about removing this from our build and installation.

I did however find the `MagickCMD.cpp` file in our git history: https://github.com/ImageMagick/contrib/blob/b178d4bfaec0e086feb0f666fcb76330997c3c7c/win32/ATL7/ImageMagickObject/MagickCMD.cpp

**Reply by Stormwalker** [reply]
Hi Dirk,

That's great - many thanks! 

(It's a client of ours that's using the object - I'll pass on your comments)

Thanks again

Carl.

**Reply by mwolfe02** [reply]
@dlemstra I see that you did in fact remove `ImageMagickObject` from the Windows build.  Was there any particular reason for doing so?

As far as I can tell, it was simply a thin, COM-compatible wrapper around the same functionality available via the command line.  Since all it seemed to be doing was passing the information along, it seems like any updates made to the core ImageMagick code would have essentially flowed through to the COM object.  Or am I missing something?

Is there any chance you would reconsider exposing the COM object in future builds?  I'm trying to decide whether I need to start rewriting my applications to use a different approach.  

Thanks for your consideration!
Mike
